File: keyless-python/traitkeyless/keyless.py
"""keyless.py

This module provides functionality for encoding and decoding keyless addresses.

A keyless address is a type of address that doesn't rely on cryptographic key pairs for identification.
Instead, it is derived from a combination of identifiers and checksums, making it suitable for various
use cases such as tracking and verification.

Examples:
    # Encode an AppAgent keyless address
    app_agent_id = 123
    app_agent_address = encode_app_agent_address(app_agent_id)

    # Decode the keyless address
    decoded_app_agent_id = decode_app_agent_address(app_agent_address)
    assert decoded_app_agent_id == app_agent_id
"""

from dataclasses import dataclass
from enum import Enum
from hashlib import blake2b
from typing import TypeAlias

from .ss58 import ss58_decode, ss58_encode

# Constants
NAMED_ADDRESS_LENGTH = 10
SS58_FORMAT__TRAIT_ASSET_HUB = 5335


# Types of address identifiers
BlockchainAddress: TypeAlias = str
BlockchainAccountId: TypeAlias = str
AppAgentId: TypeAlias = int
TransactionalAddressId: TypeAlias = int
AddressName: TypeAlias = str
SS58Format: TypeAlias = int


class AddressType(Enum):
    Regular = 0
    AppAgent = 1
    Transactional = 2
    Named = 3


@dataclass
class BlockchainAddressInfo:
    address: BlockchainAddress
    account_id: BlockchainAccountId
    address_type: AddressType
    app_agent_id: AppAgentId | None
    ta_id: TransactionalAddressId | None
    address_name: AddressName | None

    def __eq__(self: "BlockchainAddressInfo", value: object) -> bool:
        if not isinstance(value, BlockchainAddressInfo):
            return False

        return (
            self.address == value.address
            and self.account_id == value.account_id
            and self.address_type == value.address_type
            and self.app_agent_id == value.app_agent_id
            and self.ta_id == value.ta_id
            and self.address_name == value.address_name
        )


def _blake2_256(data: bytes) -> bytes:
    """
    Helper function to calculate a 32 bytes Blake2b hash for provided data, used as key for Substrate storage items

    Parameters
    ----------
    data

    Returns
    -------

    """
    return blake2b(data, digest_size=32).digest()


__allowed_chars = set("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-#")


def _validate_app_agent_id(app_agent_id: AppAgentId) -> None:
    """
    Validates that an AppAgent ID is a valid 32-bit unsigned integer.

    Args:
        app_agent_id (int): The AppAgent ID to validate.

    Raises:
        ValueError: If the AppAgent ID is not a valid 32-bit unsigned integer.
    """
    if app_agent_id < 0 or app_agent_id > 4_294_967_295:
        msg = "AppAgent ID must be unsigned 32-bit integer"
        raise ValueError(msg)


def _validate_blockchain_address(address: BlockchainAddress) -> None:
    """
    Validates that an address is a valid blockchain address.

    Args:
        address (str): The blockchain address to validate.

    Raises:
        ValueError: If the address is invalid (wrong format or length).
    """    
    if not address:
        msg = "Address cannot be empty"
        raise ValueError(msg)
    
    print(address, len(address))

    # SS58 addresses are base58 encoded and typically 49 characters long
    if not (len(address) == 49):
        msg = "Invalid address length"
        raise ValueError(msg)

    # SS58 addresses only contain base58 characters
    allowed_chars = set("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")
    if not all(c in allowed_chars for c in address):
        msg = "Address contains invalid characters"
        raise ValueError(msg)


def _validate_address_name(name: str) -> None:
    """
    Validates that a name contains only allowed ASCII characters.

    Args:
        name (str): The name to validate.

    Returns:
        bool: True if the name is valid, otherwise False.
    """
    if len(name) != NAMED_ADDRESS_LENGTH:
        msg = "Named keyless address must be of 10 chars length"
        raise ValueError(msg)

    if not all(c in __allowed_chars for c in name):
        msg = "Address name contains invalid characters"
        raise ValueError(msg)


def _encode_address(open_part: bytes, ss58_format: SS58Format) -> BlockchainAddress:
    """
    Encode an address using the given open part, open part size, and checksum size.

    Args:
        open_part (bytes): The open part of the address.
        open_part_size (int): Size of the open part.
        checksum_size (int): Size of the checksum.

    Returns:
        str: Encoded address.
    """
    # Calculate checksum using blake2_256
    checksum = _blake2_256(open_part)

    # Construct address_encoded
    address_encoded = open_part + checksum[len(open_part) :]

    return ss58_encode(address_encoded, ss58_format)


def encode_app_agent_address(
    app_agent_id: AppAgentId, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> BlockchainAddress:
    """
    Encode an AppAgent address.

    Args:
        app_agent_id (int): AppAgent ID.

    Returns:
        str: Encoded AppAgent address.
    """
    _validate_app_agent_id(app_agent_id)

    # Convert app_agent_id to little-endian bytes
    app_agent_id_bytes = app_agent_id.to_bytes(4, byteorder="little")

    # Construct open_part
    open_part = app_agent_id_bytes + bytes([AddressType.AppAgent.value])

    return _encode_address(open_part, ss58_format)


def decode_app_agent_address(
    encoded_address: BlockchainAddress, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> AppAgentId:
    """
    Decode an encoded AppAgent address.

    Args:
        encoded_address (str): Encoded AppAgent address.

    Returns:
        int: Decoded AppAgent ID.
    """
    _validate_blockchain_address(encoded_address)
    decoding_result = decode_address(encoded_address, ss58_format)

    if decoding_result.address_type is not AddressType.AppAgent:
        msg = f"Provided address is not an AppAgent address but is instance of {decoding_result.address_type}."
        raise ValueError(msg)
    if decoding_result.app_agent_id is None:
        msg = "Internal error in traitkeyless - app_agent_id of AppAgent address is None"
        raise ValueError(msg)

    return decoding_result.app_agent_id


def encode_transactional_address(
    app_agent_id: AppAgentId,
    ta_id: TransactionalAddressId,
    ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB,
) -> BlockchainAddress:
    """
    Encode a Transactional address.

    Args:
        app_agent_id (int): AppAgent ID.
        ta_id (int): Transactional address ID.

    Returns:
        str: Encoded Transactional address.
    """
    _validate_app_agent_id(app_agent_id)
    # Convert app_agent_id and ta_id to little-endian bytes
    app_agent_id_bytes = app_agent_id.to_bytes(4, byteorder="little")
    ta_id_bytes = ta_id.to_bytes(4, byteorder="little")

    # Construct open_part
    open_part = app_agent_id_bytes + bytes([AddressType.Transactional.value]) + ta_id_bytes

    return _encode_address(open_part, ss58_format)


def decode_transactional_address(
    encoded_address: BlockchainAddress, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> tuple[AppAgentId, TransactionalAddressId]:
    """
    Decode an encoded Transactional address.

    Args:
        encoded_address (str): Encoded Transactional address.

    Returns:
        Tuple[int, int]: Decoded AppAgent ID and Transactional address ID.
    """
    _validate_blockchain_address(encoded_address)
    decoding_result = decode_address(encoded_address, ss58_format)

    if decoding_result.address_type is not AddressType.Transactional:
        msg = f"Provided address is not a Transactional address but is instance of {decoding_result.address_type}."
        raise ValueError(msg)
    if decoding_result.app_agent_id is None:
        msg = "Internal error in traitkeyless - app_agent_id of Transactional address is None"
        raise ValueError(msg)
    if decoding_result.ta_id is None:
        msg = "Internal error in traitkeyless - ta_id of Transactional address is None"
        raise ValueError(msg)

    return decoding_result.app_agent_id, decoding_result.ta_id


def encode_named_address(
    app_agent_id: AppAgentId, name: AddressName, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> BlockchainAddress:
    """
    Encode a Named address.

    Args:
        app_agent_id (int): AppAgent ID.
        name (str): Address name.

    Returns:
        str: Encoded Named address.
    """
    _validate_address_name(name)

    # Convert app_agent_id to little-endian bytes
    app_agent_id_bytes = app_agent_id.to_bytes(4, byteorder="little")

    # Construct open_part
    open_part = app_agent_id_bytes + bytes([AddressType.Named.value]) + name.encode()

    return _encode_address(open_part, ss58_format)


def decode_named_address(
    encoded_address: BlockchainAddress, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> tuple[AppAgentId, AddressName]:
    """
    Decode an encoded Named address.

    Args:
        encoded_address (str): Encoded Named address.

    Returns:
        Tuple[int, str]: Decoded AppAgent ID and address name.
    """
    _validate_blockchain_address(encoded_address)
    decoding_result = decode_address(encoded_address, ss58_format)

    if decoding_result.address_type is not AddressType.Named:
        msg = f"Provided address is not a Named address but is instance of {decoding_result.address_type}."
        raise ValueError(msg)
    if decoding_result.app_agent_id is None:
        msg = "Internal error in traitkeyless - app_agent_id of Named address is None"
        raise ValueError(msg)
    if decoding_result.address_name is None:
        msg = "Internal error in traitkeyless - name of Named address is None"
        raise ValueError(msg)

    return decoding_result.app_agent_id, decoding_result.address_name


def decode_address(
    blockchain_address: BlockchainAddress, ss58_format: SS58Format = SS58_FORMAT__TRAIT_ASSET_HUB
) -> BlockchainAddressInfo:
    """
    Decode an encoded blockchain address.

    Args:
        encoded_address (str): Encoded keyless address of any type.

    Returns:
        an object with info about the address
    """
    _validate_blockchain_address(blockchain_address)
    # Decode the encoded address
    account_id = ss58_decode(blockchain_address, ss58_format)
    account_id_bytes = bytes.fromhex(account_id[2:])

    # Read the byte that encodes type of the address
    address_type_byte = account_id_bytes[4]

    if address_type_byte == AddressType.AppAgent.value:
        # Verify checksum
        open_part_size = 5
        open_part = account_id_bytes[:open_part_size]
        checksum = account_id_bytes[open_part_size:]
        checksum_calculated = _blake2_256(open_part)[open_part_size:]
        if checksum == checksum_calculated:
            # Extract and decode app_agent_id
            app_agent_id_bytes = account_id_bytes[:4]
            app_agent_id = int.from_bytes(app_agent_id_bytes, byteorder="little")

            return BlockchainAddressInfo(
                address=blockchain_address,
                account_id=account_id,
                address_type=AddressType.AppAgent,
                app_agent_id=app_agent_id,
                ta_id=None,
                address_name=None,
            )

    if address_type_byte == AddressType.Transactional.value:
        # Verify checksum
        open_part_size = 9
        open_part = account_id_bytes[:open_part_size]
        checksum = account_id_bytes[open_part_size:]
        checksum_calculated = _blake2_256(open_part)[open_part_size:]
        if checksum == checksum_calculated:
            # Extract and decode app_agent_id and ta_id
            app_agent_id_bytes = account_id_bytes[:4]
            ta_id_bytes = account_id_bytes[5:9]
            app_agent_id = int.from_bytes(app_agent_id_bytes, byteorder="little")
            ta_id = int.from_bytes(ta_id_bytes, byteorder="little")

            return BlockchainAddressInfo(
                address=blockchain_address,
                account_id=account_id,
                address_type=AddressType.Transactional,
                app_agent_id=app_agent_id,
                ta_id=ta_id,
                address_name=None,
            )

    if address_type_byte == AddressType.Named.value:
        # Verify checksum
        open_part_size = 15
        open_part = account_id_bytes[:open_part_size]
        checksum = account_id_bytes[open_part_size:]
        checksum_calculated = _blake2_256(open_part)[open_part_size:]
        if checksum == checksum_calculated:
            # Extract and decode app_agent_id and address_name
            app_agent_id_bytes = account_id_bytes[:4]
            name_bytes = account_id_bytes[5:15]
            app_agent_id = int.from_bytes(app_agent_id_bytes, byteorder="little")
            address_name = name_bytes.decode()

            return BlockchainAddressInfo(
                address=blockchain_address,
                account_id=account_id,
                address_type=AddressType.Named,
                app_agent_id=app_agent_id,
                ta_id=None,
                address_name=address_name,
            )

    return BlockchainAddressInfo(
        address=blockchain_address,
        account_id=account_id,
        address_type=AddressType.Regular,
        app_agent_id=None,
        ta_id=None,
        address_name=None,
    )

File: keyless-ts/src/keyless.ts
import { blake2AsU8a } from "@polkadot/util-crypto";
import {
  decodeAddress as decodeAddressSs58,
  encodeAddress,
} from "@polkadot/keyring";
import { strict as assert } from "assert";

const APP_AGENT_ADDRESS_IDENTIFIER = 1;
const TRANSACTIONAL_ADDRESS_IDENTIFIER = 2;
const NAMED_ADDRESS_IDENTIFIER = 3;
const NAMED_ADDRESS_BYTES_LENGTH = 10;
const SS58_FORMAT__TRAIT_ASSET_HUB = 5335;

enum AddressType {
  Regular = 0,
  AppAgent = 1,
  Transactional = 2,
  Named = 3,
}

// Represents the information related to a blockchain address.
interface BlockchainAddressInfo {
  address: string;
  accountId: string;
  addressType: AddressType;
  appAgentId: number | null;
  taId: number | null;
  addressName: string | null;
}

/**
 * Validates that a name contains only allowed ASCII characters.
 *
 * @param {string} name - The name to validate.
 * @returns {boolean} - True if the name is valid
 */
function validateAddressName(name: string): void {
  assert(name.length === NAMED_ADDRESS_BYTES_LENGTH, "Name size not supported");
  const allowedChars = /^[0-9a-zA-Z\-#]+$/;
  assert(allowedChars.test(name), "Name contains invalid characters");
}

/**
 * Encodes an application agent address and returns it.
 *
 * @param {number} appAgentId - The ID of the application agent.
 * @returns {string} - The encoded application agent address.
 */
function encodeAppAgent(
  appAgentId: number,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): string {
  const appAgentIdBytes = Buffer.alloc(4);
  appAgentIdBytes.writeUInt32LE(appAgentId, 0);

  const appAgentIdBuffer = Buffer.from([APP_AGENT_ADDRESS_IDENTIFIER]);
  const openPart = Buffer.concat([appAgentIdBytes, appAgentIdBuffer]);
  const checksum = blake2AsU8a(openPart);
  const address = Buffer.concat([openPart, checksum.slice(openPart.length)]);

  return encodeAddress(address, ss58Format);
}

/**
 * Encodes a transactional agent address and returns it.
 *
 * @param {number} appAgentId - The ID of the application agent.
 * @param {number} taId - The ID of the transactional agent.
 * @returns {string} - The encoded transactional agent address.
 */
function encodeTransactional(
  appAgentId: number,
  taId: number,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): string {
  const appAgentIdBytes = Buffer.alloc(4);
  appAgentIdBytes.writeUInt32LE(appAgentId, 0);

  const taIdBytes = Buffer.alloc(4);
  taIdBytes.writeUInt32LE(taId, 0);

  const taIdBuffer = Buffer.from([TRANSACTIONAL_ADDRESS_IDENTIFIER]);
  const openPart = Buffer.concat([appAgentIdBytes, taIdBuffer, taIdBytes]);
  const checksum = blake2AsU8a(openPart);
  const address = Buffer.concat([openPart, checksum.slice(openPart.length)]);

  return encodeAddress(address, ss58Format);
}

/**
 * Encodes a named address and returns it.
 *
 * @param {number} appAgentId - The ID of the application agent.
 * @param {string} name - The name of the address.
 * @returns {string} - The encoded named address.
 */
function encodeNamed(
  appAgentId: number,
  name: string,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): string {
  validateAddressName(name);

  const appAgentIdBytes = Buffer.alloc(4);
  appAgentIdBytes.writeUInt32LE(appAgentId, 0);

  const nameBytes = Buffer.from(name);

  const taIdBuffer = Buffer.from([NAMED_ADDRESS_IDENTIFIER]);
  const openPart = Buffer.concat([appAgentIdBytes, taIdBuffer, nameBytes]);
  const checksum = blake2AsU8a(openPart);
  const address = Buffer.concat([openPart, checksum.slice(openPart.length)]);

  return encodeAddress(address, ss58Format);
}

function isEqualArrays(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

/**
 * Decodes a keyless address and determines its type.
 *
 * @param {string} address - The blockchain address to decode.
 * @returns {BlockchainAddressInfo} - An object containing the decoded address type and data.
 * @throws {Error} - Thrown if the address cannot be decoded or is of an unsupported type.
 */
function decodeAddress(
  address: string,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): BlockchainAddressInfo {
  const accountId = decodeAddressSs58(address, false, ss58Format);
  const accountIdBuffer = Buffer.from(accountId);

  // Read the byte that encodes type of the address
  const addressTypeByte = accountId[4];

  if (addressTypeByte === APP_AGENT_ADDRESS_IDENTIFIER) {
    // Verify checksum
    const openPartSize = 5;
    const openPart = accountId.slice(0, openPartSize);
    const checksum = accountId.slice(openPartSize);
    const checksumCalculated = blake2AsU8a(openPart).slice(openPartSize);
    if (isEqualArrays(checksum, checksumCalculated)) {
      // Extract and decode appAgentId
      const appAgentId = accountIdBuffer.subarray(0, 4).readUint32LE(0);

      return {
        address,
        accountId: `0x${accountIdBuffer.toString("hex")}`,
        addressType: AddressType.AppAgent,
        appAgentId,
        taId: null,
        addressName: null,
      };
    }
  }

  if (addressTypeByte === TRANSACTIONAL_ADDRESS_IDENTIFIER) {
    // Verify checksum
    const openPartSize = 9;
    const openPart = accountId.slice(0, openPartSize);
    const checksum = accountId.slice(openPartSize);
    const checksumCalculated = blake2AsU8a(openPart).slice(openPartSize);
    if (isEqualArrays(checksum, checksumCalculated)) {
      // Extract and decode appAgentId_bytes and taId_bytes
      const appAgentId = accountIdBuffer.subarray(0, 4).readUint32LE(0);
      const taId = accountIdBuffer.subarray(5, 9).readUint32LE(0);

      return {
        address,
        accountId: `0x${accountIdBuffer.toString("hex")}`,
        addressType: AddressType.Transactional,
        appAgentId,
        taId,
        addressName: null,
      };
    }
  }

  if (addressTypeByte === NAMED_ADDRESS_IDENTIFIER) {
    // Verify checksum
    const openPartSize = 15;
    const openPart = accountId.slice(0, openPartSize);
    const checksum = accountId.slice(openPartSize);
    const checksumCalculated = blake2AsU8a(openPart).slice(openPartSize);
    if (isEqualArrays(checksum, checksumCalculated)) {
      // Extract and decode appAgentId and address_name
      const appAgentId = accountIdBuffer.subarray(0, 4).readUint32LE(0);
      const addressName = accountIdBuffer.subarray(5, 15).toString();

      return {
        address,
        accountId: `0x${accountIdBuffer.toString("hex")}`,
        addressType: AddressType.Named,
        appAgentId,
        taId: null,
        addressName,
      };
    }
  }

  return {
    address,
    accountId: `0x${accountIdBuffer.toString("hex")}`,
    addressType: AddressType.Regular,
    appAgentId: null,
    taId: null,
    addressName: null,
  };
}

/**
 * Decodes an application agent address.
 *
 * @param {string} address
 * @returns {number} - The ID of the application agent.
 */
function decodeAppAgent(
  address: string,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): number {
  const decodingResult = decodeAddress(address, ss58Format);

  if (decodingResult.addressType !== AddressType.AppAgent) {
    throw new Error(
      `Provided address is not an AppAgent address but is instance of ${AddressType[decodingResult.addressType]}`,
    );
  }
  if (
    decodingResult.appAgentId === undefined ||
    decodingResult.appAgentId === null
  ) {
    throw new Error(
      "Internal error in traitkeyless - appAgentId of AppAgent address is undefined",
    );
  }

  return decodingResult.appAgentId;
}

interface TransactionalAddressId {
  appAgentId: number;
  taId: number;
}

/**
 * Decodes a transactional address.
 *
 * @param {string} address - The JSON representation of the encoded transactional address.
 * @returns {Object} - An object containing the IDs of the app agent and transactional agent.
 */
function decodeTransactional(
  address: string,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): TransactionalAddressId {
  const decodingResult = decodeAddress(address, ss58Format);

  if (decodingResult.addressType !== AddressType.Transactional) {
    throw new Error(
      `Provided address is not a Transactional address but is instance of ${AddressType[decodingResult.addressType]}`,
    );
  }
  if (
    decodingResult.appAgentId === undefined ||
    decodingResult.appAgentId === null
  ) {
    throw new Error(
      "Internal error in traitkeyless - appAgentId of Transactional address is undefined",
    );
  }
  if (decodingResult.taId === undefined || decodingResult.taId === null) {
    throw new Error(
      "Internal error in traitkeyless - taId of Transactional address is undefined",
    );
  }

  return {
    appAgentId: decodingResult.appAgentId,
    taId: decodingResult.taId,
  };
}

interface NamedAddressId {
  appAgentId: number;
  addressName: string;
}

/**
 * Decodes a named address.
 *
 * @param {string} address - The JSON representation of the encoded named address.
 * @returns {BlockchainAddressInfo} - An object containing the IDs of the application agent and name.
 */
function decodeNamed(
  address: string,
  ss58Format: number = SS58_FORMAT__TRAIT_ASSET_HUB,
): NamedAddressId {
  const decodingResult = decodeAddress(address, ss58Format);

  if (decodingResult.addressType !== AddressType.Named) {
    throw new Error(
      `Provided address is not a Named address but is instance of ${AddressType[decodingResult.addressType]}`,
    );
  }
  if (
    decodingResult.appAgentId === undefined ||
    decodingResult.appAgentId === null
  ) {
    throw new Error(
      "Internal error in traitkeyless - appAgentId of Named address is undefined",
    );
  }
  if (
    decodingResult.addressName === undefined ||
    decodingResult.addressName === null
  ) {
    throw new Error(
      "Internal error in traitkeyless - addressName of Named address is undefined",
    );
  }

  return {
    appAgentId: decodingResult.appAgentId,
    addressName: decodingResult.addressName,
  };
}

export {
  AddressType,
  BlockchainAddressInfo,
  TransactionalAddressId,
  NamedAddressId,
  encodeAppAgent,
  encodeTransactional,
  encodeNamed,
  decodeAppAgent,
  decodeTransactional,
  decodeNamed,
  decodeAddress,
  SS58_FORMAT__TRAIT_ASSET_HUB,
};
